proxylib/proxylib/policymap.go

```diff
@@ -0,0 +1,332 @@
+}
+
+func newPortNetworkPolicies(config []*cilium.PortNetworkPolicy) PortNetworkPolicies {
+	policy := PortNetworkPolicies{
+		Rules: make(map[uint32]PortNetworkPolicyRules, len(config)),
+	}
+	for _, rule := range config {
+		// Ignore UDP policies
+		if rule.GetProtocol() == core.SocketAddress_UDP {
+			continue
+		}
+
+		port := rule.GetPort()
+		if _, found := policy.Rules[port]; found {
+			panic(fmt.Errorf("NPDS: Duplicate port number %d in (rule: %v) (config: %v)", port, rule, config))
+		}
+
+		if rule.GetProtocol() != core.SocketAddress_TCP {
+			panic(fmt.Errorf("NPDS: Invalid transport protocol %v", rule.GetProtocol()))
+		}
+
+		// Skip the port if not 'ok'
+		rules, ok := newPortNetworkPolicyRules(rule.GetRules())
+		if ok {
+			log.Debugf("NPDS::PortNetworkPolicies(): installed TCP policy for port %d", port)
+			policy.Rules[port] = rules
+		} else {
+			log.Debugf("NPDS::PortNetworkPolicies(): Skipped port due to unsupported L7: %d", port)
+		}
+	}
+	return policy
+}
+
+func (p *PortNetworkPolicies) Matches(port, remoteId uint32, l7 interface{}) bool {
+	rules, found := p.Rules[port]
+	if found {
+		if rules.Matches(remoteId, l7) {
+			log.Debugf("NPDS::PortNetworkPolicies(port=%d, remoteId=%d): rule matches (%v)", port, remoteId, p)
+			return true
+		}
+	}
+	// No exact port match, try wildcard
+	rules, foundWc := p.Rules[0]
+	if foundWc {
+		if rules.Matches(remoteId, l7) {
+			log.Debugf("NPDS::PortNetworkPolicies(port=*, remoteId=%d): rule matches (%v)", remoteId, p)
+			return true
+		}
+	}
+
+	// No policy for the port was found. Cilium always creates a policy for redirects it
+	// creates, so the host proxy never gets here. Sidecar gets all the traffic, which we need
+	// to pass through since the bpf datapath already allowed it.
+	// TODO: Change back to false only when non-bpf datapath is supported?
+
+	//	log.Debugf("NPDS::PortNetworkPolicies(port=%d, remoteId=%d): allowing traffic on port for which there is no policy, assuming L3/L4 has passed it! (%v)", port, remoteId, p)
+	//	return !(found || foundWc)
+	if !(found || foundWc) {
+		log.Debugf("NPDS::PortNetworkPolicies(port=%d, remoteId=%d): Dropping traffic on port for which there is no policy! (%v)", port, remoteId, p)
+	}
+	return false
+}
+
+type PolicyInstance struct {
+	protobuf cilium.NetworkPolicy
+	Ingress  PortNetworkPolicies
+	Egress   PortNetworkPolicies
+}
+
+func newPolicyInstance(config *cilium.NetworkPolicy) *PolicyInstance {
+	log.Debugf("NPDS::PolicyInstance: Inserting policy %s", config.String())
+	
+	return &PolicyInstance{
+		protobuf: *config,
+		Ingress:  newPortNetworkPolicies(config.GetIngressPerPortPolicies()),
+		Egress:   newPortNetworkPolicies(config.GetEgressPerPortPolicies()),
+	}
+}
+
+func (p *PolicyInstance) Matches(ingress bool, port, remoteId uint32, l7 interface{}) bool {
+	if ingress {
+		return p.Ingress.Matches(port, remoteId, l7)
+	}
+	return p.Egress.Matches(port, remoteId, l7)
+}
+
+type PolicyMap map[string]*PolicyInstance
```

Comment:
neewds comment to what is the key of this map

Reply:
Added.